clients/src/test/java/org/apache/kafka/clients/consumer/internals/NetworkClientDelegateTest.java

```diff
@@ -97,6 +99,29 @@ public void testTimeoutAfterSend() throws Exception {
         }
     }
 
+    @Test
+    public void testEnsureCorrectCompletionTimeOnFailure() {
+        NetworkClientDelegate.UnsentRequest unsentRequest = newUnsentFindCoordinatorRequest();
+        long timeMs = time.milliseconds();
+        unsentRequest.handler().onFailure(timeMs, new TimeoutException());
+
+        time.sleep(100);
+        assertEquals(timeMs, unsentRequest.handler().completionTimeMs());
+    }
+
+    @Test
+    public void testEnsureCorrectCompletionTimeOnComplete() throws IOException {
+        NetworkClientDelegate.UnsentRequest unsentRequest = newUnsentFindCoordinatorRequest();
+        prepareFindCoordinatorResponse(Errors.NONE);
+        long timeMs = time.milliseconds();
+        try (NetworkClientDelegate delegate = newNetworkClientDelegate()) {
+            delegate.send(unsentRequest);
+            delegate.poll(0, timeMs);
+        }
+        time.sleep(100);
+        assertEquals(timeMs, unsentRequest.handler().completionTimeMs());
+    }
+
```

Comment:
If you use a mock for the client response, the test becomes simpler.
```suggestion
@Test
public void testEnsureCorrectCompletionTimeOnComplete() {
    NetworkClientDelegate.UnsentRequest unsentRequest = newUnsentFindCoordinatorRequest();
    long timeMs = time.milliseconds();
    final ClientResponse response = mock(ClientResponse.class);
    when(response.receivedTimeMs()).thenReturn(timeMs);

    unsentRequest.handler().onComplete(response);
    time.sleep(100);

    assertEquals(timeMs, unsentRequest.handler().completionTimeMs());
}
```